### PR TITLE
Implementing sharelink feature requested in #77

### PR DESCRIPTION
**Pull request by banool**
This is in response to #77.

Following the current command structure for `dbxcli`, the new command can be found at `dbxcli share getlink`. 

If the file / folder already has a sharelink, that link is retrieved and printed. This follows the pre-existing printing format from the sharing functions. If it doesn't have a sharelink, the link is created and printed.

There is also support for getting links for files that haven't fully uploaded yet. I wanted to functionality to be precisely like if you right click on a file and use Copy Dropbox Link. The side effect of this is I had to use the partially deprecated `create_shared_link` endpoint.

This is the first Go I've ever coded so I'm of course open to revisions.

### CODE REVIEW

File: cmd/share-createlink.go [hunk 1]
```diff
@@ -0,0 +1,147 @@
+// Copyright © 2017 Dropbox, Inc.
+// Author: Daniel Porteous
```
Comment (diwakergupta) [inline cmd/share-createlink.go]: For consistency, can you omit this line? `git blame` can always reveal authors :)

File: cmd/share-createlink.go [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+	"os/user"
+	"path"
+	"path/filepath"
+	"strings"
+
+	"github.com/dropbox/dropbox-sdk-go-unofficial/dropbox/sharing"
+	"github.com/spf13/cobra"
+)
+
+/**
```
Comment (diwakergupta) [inline cmd/share-createlink.go]: Please use `// ` to start comments. While go does support C style block comments, `//` is the norm and Javadoc-style comments are rare. More at https://golang.org/doc/effective_go.html#commentary

File: cmd/share-createlink.go [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+	dbx := sharing.New(config)
+	path, err := filepath.Abs(args[0])
+	if err != nil {
+		return
+	}
+
+	// Confirm that the file exists.
+	exists, err := exists(path)
+	if !exists || err != nil {
+		print("The file / folder specified does not exist.\n")
```
Comment (diwakergupta) [inline cmd/share-createlink.go]: Print the path here as well so users have context

File: cmd/share-createlink.go [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+		return
+	}
+
+	// Try to get a link if it already exists.
+	if getExistingLink(dbx, path) {
+		return
+	}
+
+	// print("File / folder does not yet have a sharelink, creating one...\n")
```
Comment (diwakergupta) [inline cmd/share-createlink.go]: Delete?

File: cmd/share-createlink.go [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+	getNewLink(dbx, path)
+
+	return
+}
+
+func printShareLinkUsage() {
+	fmt.Printf("Usage: %s share getlink [file / folder path]\n", os.Args[0])
+}
+
+/*
```
Comment (diwakergupta) [inline cmd/share-createlink.go]: Same for comments here and elsewhere, start with `//`

File: cmd/share-createlink.go [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+** Try to get an existing share link for a file / folder.
+** It returns true if the file / folder had a link. Otherwise it returns false.
+ */
+func getExistingLink(dbx sharing.Client, path string) bool {
```
Comment (diwakergupta) [inline cmd/share-createlink.go]: The name of this function is counter-intuitive, as it doesn't actually return the existing link. Perhaps `checkExistingLink`?

File: cmd/share-createlink.go [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+	return false
+}
+
+/*
+** Create and print a link for file / folder that doesn't yet have one.
+**
+** CreateSharedLinkWithSettings doesn't allow pending uploads,
+** so we use the partially deprecated CreateSharedLink.
+ */
+func getNewLink(dbx sharing.Client, path string) bool {
```
Comment (diwakergupta) [inline cmd/share-createlink.go]: Looks like you are using bool to handle errors. Why not return `error` instead?

File: cmd/share-createlink.go [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+	arg := sharing.NewCreateSharedLinkArg(strings.Replace(path, getDropboxFolder(), "", 1))
+	// Get the sharelink even if the file isn't fully uploaded yet.
+	arg.PendingUpload = new(sharing.PendingUploadMode)
+	// Determine whether the target is a file or folder.
+	fi, err := os.Stat(path)
```
Comment (diwakergupta) [inline cmd/share-createlink.go]: You can do this check before creating arg

File: cmd/share-createlink.go [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+	if err != nil {
+		fmt.Printf("%+v\n", err)
+		return false
+	}
+	fmt.Printf("%s\t%s\n", res.Path[1:], res.Url)
+	return true
+}
+
+/* Return the path of the Dropbox folder. */
+func getDropboxFolder() string {
```
Comment (diwakergupta) [inline cmd/share-createlink.go]: This whole function is super hacky and likely not cross-platform. Let me think if there's a better way.

**Review by diwakergupta — CHANGES_REQUESTED** [review]

File: cmd/share-createlink.go (hunk 6 above)
Comment (banool) [reply cmd/share-createlink.go]: It doesn't just check the existing link though, it also returns it if it exists. I think `checkExistingLinkAndReturnIfExists` would be a bit wordy 😉. I can change it regardless if you want.

File: cmd/share-createlink.go (hunk 7 above)
Comment (banool) [reply cmd/share-createlink.go]: Sounds like a good idea, I'll try that.

File: cmd/share-createlink.go (hunk 9 above)
Comment (banool) [reply cmd/share-createlink.go]: Awesome, thanks. I'll address the other issues now and we can come back to this when you have a solution. There should be a function somewhere that returns where the Dropbox folder is, the information has to be held somewhere.

**Comment by banool** [comment]
I addressed all of the issues except those related to `getDropboxFolder`.

**Comment by coatless** [comment]
Any ETA on when this feature can be live?

**Comment by banool** [comment]
@coatless Just waiting on the go ahead from @diwakergupta, as well as perhaps a cleaner way to do the functionality provided by `getDropboxFolder`.

**Comment by coatless** [comment]
@banool @diwakergupta it would be great to see headway on this feature as the `~/dropbox.py` script appears to be at EOL given:

```bash
root@remote:~# ~/dropbox.py status
You're using an old version of Dropbox. Please update within the next 4 days to continue using Dropbox.

https://www.dropbox.com/downloading?from_client=True
Up to date
```

**Comment by diwakergupta** [comment]
Sorry folks, will take a look right away!

File: cmd/share-createlink.go (hunk 9 above)
Comment (diwakergupta) [reply cmd/share-createlink.go]: Given the complexity around obtaining the "canonical" Dropbox folder, for all scenarios (e.g. paid users, paired accounts, custom directories) and for all platforms, I'd prefer if we solve this via documentation and help, and being explicit. For instance, we can ask users to just supply the path instead of trying to prefix things ourselves. WDYT?

**Comment by MartinDelille** [comment]
@diwakergupta Any update on this PR? I'd love to use this command.

I installed `dbxcli` only for this purpose because my current workflow is:

- `open -r myfile`
- Right click on the file: copy shareable link

**Comment by MartinDelille** [comment]
@banool is there a way to install your version to test it?

**Comment by cahebert** [comment]
Bumping this, I would love to see this feature incorporated into the main branch.

**Comment by batt** [comment]
I would love to have this PR merged too!

**Comment by vlaminck** [comment]
@diwakergupta any update on if this will get merged? It would be really nice to have.

**Comment by vlaminck** [comment]
I know this is super old and seems to not actually be maintained any more, but when I try to `go build` your branch, I get this error:

```
./main.go:43:24: cannot use versionCmd (type *"github.com/banool/dbxcli/vendor/github.com/spf13/cobra".Command) as type *"github.com/dropbox/dbxcli/vendor/github.com/spf13/cobra".Command in argument to cmd.RootCmd.AddCommand
```

**Comment by coatless** [comment]
Sadly, I think this software is official dead. I wouldn't push forward more on this as it literally was a year and then some. Consider looking into Dropbox alternatives.

**Comment by banool** [comment]
Unfortunately yes, I haven't looked into this PR for a long time. I don't intend to pick up the work again either since, as @coatless says, this appears to be entirely dead. Dropbox seems to be more interested in the designer / creative space than the technical space these days, so I don't expect to see any further development on this either. Given the recent price increase and worsening features relative to competitors, might be time to look elsewhere.

**Comment by vlaminck** [comment]
Yeah, I figured that'd be the case. Thanks for the responses.

If anyone finds this in the future, here's a Groovy script that I wrote to accomplish what I needed. https://gitlab.com/snippets/1915960

**Comment by MartinDelille** [comment]
I wanted to give it a look and I manage to make it work!

I rebased the PR (otherwise I had a segmentation fault).

```
$ dbxcli share getlink ~/Dropbox/Flyers\ Lyon.jpg
2019/11/23 19:38:45 client.go:619: WARNING: API `CreateSharedLink` is deprecated
2019/11/23 19:38:45 client.go:620: Use API `CreateSharedLinkWithSettings` instead
Flyers Lyon.jpg	https://www.dropbox.com/s/1mvz1bhiq7csq1l/Flyers%20Lyon.jpg?dl=0
```

As you can see, `CreateSharedLink` is deprecated but it shouldn't be a problem to fix that.

Regarding the output I would remove the file name in order to make the result  "pipable":

```
$ dbxcli share getlink ~/Dropbox/Flyers\ Lyon.jpg | pbcopy
```

(for non *MacOS* user, `pbcopy` copy the input in the clipboard)

**Comment by biounix** [comment]
https://github.com/dropbox/dbxcli/issues/113#issuecomment-667919621

File: cmd/share-createlink.go (hunk 9 above)
Comment (arafatx) [reply cmd/share-createlink.go:111]: > pbcopy


Is this available in PR ?

**Comment by banool** [comment]
I'm unsubscribing from this, please tag me if you need me if the maintainers ever return. Thanks!

**Comment by MartinDelille** [comment]
@banool If you want to get notified, stay subscribed because I don't think I will remember to ping you when it merged (there is not so much activity here anyway).

@biounix I just tried tbx for the first time after having it under my watched repository since you mentioned it. I opened [a question here](https://github.com/watermint/toolbox/issues/594) about it.

**Comment by AndreyVMarkelov** [comment]
Thanks for the PR and for the original share-link work. This has been superseded by the current `share-link` command family on `master`.

Current `dbxcli` now supports:

- `dbxcli share-link create`
- `dbxcli share-link list`
- `dbxcli share-link info`
- `dbxcli share-link update`
- `dbxcli share-link revoke`
- `dbxcli share-link download`

`share-link create` creates a new link or returns an existing direct link, and the newer implementation includes structured JSON output, tests, docs, and support for link settings such as audience, expiration, password, and download permissions.

Given that this PR is based on the old `share getlink` shape and has been replaced by the newer implementation, I am closing it as obsolete.
